From ab101021de9c3881cf70d9fc5e1af2dc2d38ebf9 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 10 Jan 2017 16:08:03 +0800
Subject: user32: Add support for navigating a group of radio buttons using a
 keyboard.

It approximates the behaviour observed in the message tests but still
doesn't make the message tests pass without failures.
---
 dlls/user32/dialog.c    | 12 ++++++++++--
 dlls/user32/tests/msg.c |  7 -------
 2 files changed, 10 insertions(+), 9 deletions(-)

diff --git a/dlls/user32/dialog.c b/dlls/user32/dialog.c
index 7f9a72d59ee..f8571498643 100644
--- a/dlls/user32/dialog.c
+++ b/dlls/user32/dialog.c
@@ -1239,8 +1239,16 @@ BOOL WINAPI IsDialogMessageW( HWND hwndDlg, LPMSG msg )
             if (!(dlgCode & DLGC_WANTARROWS))
             {
                 BOOL fPrevious = (msg->wParam == VK_LEFT || msg->wParam == VK_UP);
-                HWND hwndNext = GetNextDlgGroupItem (hwndDlg, GetFocus(), fPrevious );
-                SendMessageW( hwndDlg, WM_NEXTDLGCTL, (WPARAM)hwndNext, 1 );
+                HWND hwndNext = GetNextDlgGroupItem( hwndDlg, msg->hwnd, fPrevious );
+                if (hwndNext && SendMessageW( hwndNext, WM_GETDLGCODE, msg->wParam, (LPARAM)msg ) == (DLGC_BUTTON | DLGC_RADIOBUTTON))
+                {
+                    SetFocus( hwndNext );
+                    if ((GetWindowLongW( hwndNext, GWL_STYLE ) & BS_TYPEMASK) == BS_AUTORADIOBUTTON &&
+                        SendMessageW( hwndNext, BM_GETCHECK, 0, 0 ) != BST_CHECKED)
+                        SendMessageW( hwndNext, BM_CLICK, 1, 0 );
+                }
+                else
+                    SendMessageW( hwndDlg, WM_NEXTDLGCTL, (WPARAM)hwndNext, 1 );
                 return TRUE;
             }
             break;
diff --git a/dlls/user32/tests/msg.c b/dlls/user32/tests/msg.c
index ba7320d9c41..bfd2b915280 100644
--- a/dlls/user32/tests/msg.c
+++ b/dlls/user32/tests/msg.c
@@ -6451,11 +6451,9 @@ if (0) /* actual message sequence is different on every run in some Windows setu
     while (PeekMessageA(&msg, 0, 0, 0, PM_REMOVE)) DispatchMessageA(&msg);
     ok_sequence(auto_radio_button_VK_DOWN_dialog, "IsDialogMessage(VK_DOWN)", TRUE);
 
-todo_wine
     test_radio(radio1, 1, radio2, 0, radio3, 0);
 
     hwnd = GetFocus();
-todo_wine
     ok(hwnd == radio1, "focus should be on radio1, not on %p\n", hwnd);
     GetWindowRect(radio1, &rc);
 
@@ -6470,11 +6468,9 @@ todo_wine
     while (PeekMessageA(&msg, 0, 0, 0, PM_REMOVE)) DispatchMessageA(&msg);
     ok_sequence(auto_radio_button_VK_DOWN_radio3, "down to radio3", TRUE);
 
-todo_wine
     test_radio(radio1, 1, radio2, 0, radio3, 0);
 
     hwnd = GetFocus();
-todo_wine
     ok(hwnd == radio1, "focus should be on radio1, not on %p\n", hwnd);
 
     flush_events();
@@ -6491,11 +6487,9 @@ todo_wine
     while (PeekMessageA(&msg, 0, 0, 0, PM_REMOVE)) DispatchMessageA(&msg);
     ok_sequence(auto_radio_button_VK_UP_radio1, "up to radio1", TRUE);
 
-todo_wine
     test_radio(radio1, 1, radio2, 0, radio3, 0);
 
     hwnd = GetFocus();
-todo_wine
     ok(hwnd == radio1, "focus should be on radio1, not on %p\n", hwnd);
 
     flush_events();
@@ -6513,7 +6507,6 @@ todo_wine
 if (0) /* actual message sequence is different on every run in some Windows setups */
     ok_sequence(auto_radio_button_VK_UP_dialog, "IsDialogMessage(VK_UP) #3", FALSE);
     /* what really matters is that nothing has changed */
-todo_wine
     test_radio(radio1, 1, radio2, 0, radio3, 0);
 
     log_all_parent_messages--;
-- 
2.11.0

